--- linux/net/Config.in	Mon May 20 12:15:38 2002
+++ ebt2.0pre6/net/Config.in	Mon May 20 11:58:15 2002
@@ -60,7 +60,9 @@
    source net/decnet/Config.in
 fi
 dep_tristate '802.1d Ethernet Bridging' CONFIG_BRIDGE $CONFIG_INET
+if [ "$CONFIG_BRIDGE" != "n" -a "$CONFIG_NETFILTER" != "n" ]; then
    source net/bridge/netfilter/Config.in
+fi
 if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
    if [ "$CONFIG_BRIDGE" != "n" -a "$CONFIG_NETFILTER" != "n" ]; then
       bool '  netfilter (firewalling) support' CONFIG_BRIDGE_NF
