--- linux-2.4.20-pre7/net/bridge/br_input.c	Sun Sep 29 21:34:42 2002
+++ linux-2.4.20-pre7-ebt/net/bridge/br_input.c	Sun Sep 29 21:24:37 2002
@@ -149,8 +149,10 @@
 		goto handle_special_frame;
 
 	if (p->state == BR_STATE_FORWARDING) {
-		if (br_should_route_hook && br_should_route_hook(&skb))
+		if (br_should_route_hook && br_should_route_hook(&skb)) {
+			read_unlock(&br->lock);
 			return -1;
+		}
 
 		NF_HOOK(PF_BRIDGE, NF_BR_PRE_ROUTING, skb, skb->dev, NULL,
 			br_handle_frame_finish);
