--- ebtables-v2.0pre2.002/Makefile	Sat Apr 13 17:39:08 2002
+++ ebtables-v2.0pre2.003/Makefile	Sat Apr 13 17:43:09 2002
@@ -2,7 +2,7 @@
 
 KERNEL_DIR?=/usr/src/linux
 PROGNAME:=ebtables
-PROGVERSION:="2.0pre2.001 (April 2002)"
+PROGVERSION:="2.0pre2.003 (April 2002)"
 
 MANDIR?=/usr/local/man
 CFLAGS:=-Wall -Wunused
--- ebtables-v2.0pre2.002/ebtables.c	Sat Apr 13 17:39:08 2002
+++ ebtables-v2.0pre2.003/ebtables.c	Sat Apr 13 17:37:00 2002
@@ -1490,7 +1490,7 @@
 		flush_chains();
 	else if (replace.command == 'A' || replace.command == 'I')
 		add_rule(rule_nr);
-	else
+	else if (replace.command == 'D')
 		delete_rule(rule_nr);
 
 	if (table->check)
--- ebtables-v2.0pre2.002/THANKS	Wed Apr  3 17:44:44 2002
+++ ebtables-v2.0pre2.003/THANKS	Sat Apr 13 17:40:35 2002
@@ -5,3 +5,5 @@
 Harald Welte
 Jason Lunz
 Tim Gardner
+Loïc Minier
+
--- ebtables-v2.0pre2.002/ebtables.8	Wed Apr  3 16:29:11 2002
+++ ebtables-v2.0pre2.003/ebtables.8	Sat Apr 13 17:32:23 2002
@@ -128,6 +128,7 @@
 Put the counters of the selected chain on zero. If no chain is selected, all the counters
 are put on zero. This can be used in conjunction with the -L command (see above). 
 This will cause the rule counters to be printed on the screen before they are put on zero.
+.TP
 .B "-P, --policy"
 Set the policy for the chain to the given target. The policy is either
 .B ACCEPT
