--- ebtables-v2.0-rc2/Makefile	Sat Aug 31 11:51:29 2002
+++ ebtables-v2.0/Makefile	Thu Sep 19 19:52:09 2002
@@ -2,11 +2,12 @@
 
 KERNEL_DIR?=/usr/src/linux
 PROGNAME:=ebtables
-PROGVERSION:="2.0-rc2"
-PROGDATE:="August 2002"
+PROGVERSION:="2.0"
+PROGDATE:="September 2002"
 
 MANDIR?=/usr/local/man
 CFLAGS:=-Wall -Wunused
+CC:=gcc
 include extensions/Makefile
 
 # Some kernel testers prefer to use a symlink for /usr/include/linux
